File: test/conformance/js/rows/F4.mjs
// F4 — layout_malformed, truncated (docs/FIXED-FORM-ALGORITHM.md §2 step 3).
//
// A fixed-form file is: the 16-byte header (form byte 3, seven reserved zero
// bytes, the layout hash at 8), then the u32 LAYOUT LENGTH L at 16, then L
// layout bytes, then the records. The layout therefore ends at `20 + L`.
// Step 3 says: `L := LE(4, b+16)`; if `20 + L > bytes`, REFUSE layout_malformed.
//
// That is the TRUNCATED case, and it is a REFUSAL BY NAME (report.refused =
// TableFixedRefusal.LayoutMalformed) and never the `malformed` FLAG. The flag
// is F7, "malformed, under 20 bytes" — a file shorter than the header — which
// is the residue and not a name. The two are distinct: this row proves the
// NAME, fired when the file reaches the header but is cut inside the layout.
//
// THE CHECK RUNS BEFORE THE HASH SELECT, so the header's hash is irrelevant to
// it: a truncated layout is refused layout_malformed even for a hash in no
// lineage, before layout_newer or no_layout could have their say.
//
// Reaches the generated tree exactly as the driver does (test/conformance/js/
// main.mjs): the same SCHEMA_JS_GENERATED indirection and the same module
// resolve, so a negative control that points the same path at a sabotaged copy
// of the generated modules makes this row red.

import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

const generated = process.env.SCHEMA_JS_GENERATED ?? "build/tables-generated-js";
const load = (path) => import(pathToFileURL(resolve(generated, path)).href);

const tables = await load("examples/TablesTable.js");
const home = await load("examples/TabledemoTable.js");

let failed = false;
function check(ok, what) {
  console.log((ok ? "ok  " : "FAIL") + " " + what);
  if (!ok) { failed = true; }
}

// The production entrypoint: WeaponConfigFixedLoad, the emitted reader. Its
// call site is this file and the law it exercises is the truncation guard in
// fixedmodule.go (`if (layoutBytes + layoutAt > byteLength)`), emitted into
// every FixedLoad. Body 22 bytes, record 30, layout 191 — the smallest fixed
// root in the tables corpus, so the numbers below are its own constants.
const HEADER = home.TableFixedHeaderBytes;          // 16
const LENGTH_AT = home.TableFixedHeaderBytes;       // the u32 L at offset 16
const LAYOUT_AT = HEADER + home.TableFixedLayoutHeaderBytes; // 20
const L = tables.WeaponConfigFixedLayoutBytes;      // the layout's own length

const one = [new home.WeaponConfig()];
const file = new Uint8Array(tables.WeaponConfigFixedMeasure(1));
check(tables.WeaponConfigFixedSave(one, 1, file) === file.length,
  "positive: a lawful file saves (length " + file.length + " = 20 + " + L + " + record 30)");

{
  const back = [new home.WeaponConfig()];
  const rep = new home.TableFixedReport();
  const n = tables.WeaponConfigFixedLoad(back, back.length, file, file.length, tables.WeaponConfigFixedNewPlan(), rep);
  check(n === 1 && rep.refused === 0 && !rep.malformed,
    "positive: the whole file reads clean (n=" + n + ", refused=" + rep.refused + ", malformed=" + rep.malformed + ")");
}

// THE TRUNCATED LAYOUT: cut the file one byte into the layout, so 20 + L > bytes.
{
  const cut = file.subarray(0, LAYOUT_AT + L - 1);
  const back = [new home.WeaponConfig()];
  const rep = new home.TableFixedReport();
  const n = tables.WeaponConfigFixedLoad(back, back.length, cut, cut.length, tables.WeaponConfigFixedNewPlan(), rep);
  check(n === -1 && rep.refused === home.TableFixedRefusal.LayoutMalformed && !rep.malformed,
    "F4: a layout cut to 20+L-1 bytes is layout_malformed, the NAME not the flag (n=" + n + ", refused=" + rep.refused + ", malformed=" + rep.malformed + ")");
}

// THE SAME NAME FOR ANY OVER-RUN LENGTH, using a minimal vector from the law:
// the check is `20 + L > bytes`, so a file of exactly 20 bytes that claims any
// nonzero L is truncated, no matter what the hash says.
{
  const v = new Uint8Array(20);
  v[0] = home.TableFixedForm;
  new DataView(v.buffer).setUint32(LENGTH_AT, 1, true); // L = 1, so 20 + 1 > 20
  const back = [new home.WeaponConfig()];
  const rep = new home.TableFixedReport();
  const n = tables.WeaponConfigFixedLoad(back, back.length, v, v.length, tables.WeaponConfigFixedNewPlan(), rep);
  check(n === -1 && rep.refused === home.TableFixedRefusal.LayoutMalformed && !rep.malformed,
    "F4: a 20-byte header claiming L=1 is layout_malformed before the hash select (refused=" + rep.refused + ")");
}

// THE BOUNDARY: `20 + L == bytes` is NOT truncated. Cut the file exactly at the
// end of the layout (drop the records), so 20 + L == bytes and rest == 0: the
// reader proceeds past the guard and reports zero records, no refusal, no flag.
{
  const exact = file.subarray(0, LAYOUT_AT + L);
  const back = [new home.WeaponConfig()];
  const rep = new home.TableFixedReport();
  const n = tables.WeaponConfigFixedLoad(back, back.length, exact, exact.length, tables.WeaponConfigFixedNewPlan(), rep);
  check(n === 0 && rep.refused === 0 && !rep.malformed,
    "F4 boundary: exactly 20+L bytes is NOT truncated — zero records, no refusal (n=" + n + ", refused=" + rep.refused + ")");
}

// THE DISTINCTION FROM F7: a file UNDER the 20-byte header is the malformed
// FLAG, never the layout_malformed NAME. The guard this row pins is the NAME,
// so a short header must NOT come back refused=LayoutMalformed.
{
  const short = file.subarray(0, 19);
  const back = [new home.WeaponConfig()];
  const rep = new home.TableFixedReport();
  const n = tables.WeaponConfigFixedLoad(back, back.length, short, short.length, tables.WeaponConfigFixedNewPlan(), rep);
  check(n === -1 && rep.malformed && rep.refused === 0,
    "F4 vs F7: a 19-byte file is the malformed FLAG, not layout_malformed (refused=" + rep.refused + ")");
}

if (failed) {
  console.log("FAILED");
  process.exit(1);
}
console.log("F4 OK");

File: test/conformance/c/rows/F4.c
/* F4 -- layout_malformed, truncated, the C leg of the schema matrix cell
   (docs/roadmap.sexp row group "file-envelope"; docs/FIXED-FORM-ALGORITHM.md
   §2 step 3, the load's third line):

       "`L := LE(4, b+16)`; if `20 + L > bytes`, `REFUSE layout_malformed`"

   A fixed-form file is the sixteen-byte header (form byte 3, seven reserved
   zero bytes, the layout hash at 8), then the u32 LAYOUT LENGTH L at 16,
   then L layout bytes, then the records -- so the layout ends at 20 + L. A
   file whose bytes stop inside the layout it announces is TRUNCATED, and the
   law's answer is the NAME layout_malformed (report->refused with
   report->reason = SCHEMA_TABLE_LAYOUT_MALFORMED) and never the `malformed`
   FLAG: the flag is F7, "under 20 bytes", the residue a name cannot be put
   on; this row is the NAME, and the two must not swap.

   THE CHECK IS STEP 3 AND THE HASH SELECT IS STEPS 5 AND 6, so the header's
   hash is irrelevant to it: a truncated layout is refused layout_malformed
   even for a hash in no lineage, where the hash select alone would have
   answered layout_newer. And the comparison is `>` and not `>=`: a file that
   ends EXACTLY at the layout's end (20 + L == bytes) is not truncated -- it
   holds zero records and reads.

   VECTOR DERIVATION. The positive file is what the generated writer itself
   lays down: `cell_fixed_save` writes the form byte, the seven reserved
   zeros, this build's own layout hash at 8, the TRUE L at 16
   (cell_fixed_layout_bytes), the layout, then one 24-byte record (8 hash +
   16 body). Every case below is that one file, or a minimal header built by
   hand, with EXACTLY ONE thing moved:

     - truncated: the caller's extent is cut one byte inside the layout,
       bytes = 20 + L - 1, so 20 + L > bytes and nothing else is touched;
     - minimal: twenty zeroed bytes, the form byte at 0, an unknown hash at
       8 (0xDEADBEEFCAFEF00D, in no lineage), L = 1 at 16 -- the smallest
       vector the law admits, claiming one more layout byte than the file
       holds, under a hash that would name layout_newer if the select ran
       first;
     - boundary: the extent stops exactly at the layout's end,
       bytes = 20 + L, the record dropped whole -- NOT truncated, n = 0;
     - under-20: bytes = 19, one byte short of the header's own twenty: the
       malformed FLAG, never the name (the F7 distinction, pinned from this
       side so the two rows cannot drift together).

   REFUSE IS TOTAL (§5.8) AND SO IS MALFORMED: on either refusal no counter
   may move and not one destination byte is written -- the destination is
   poisoned 0x5A first and checked byte for byte after.

   Build: cc -std=c11 -Wall <the -I and link flags make/c.mk gives
          build/conformance-c> test/conformance/c/rows/F4.c
          -o build/rows-c-F4 && ./build/rows-c-F4

   SELF-CONTAINED: reaches the generated fixed form through the SAME include
   path the conformance driver uses (build/tables-generated-c/v1), includes
   only V1Table.h, and depends on no other rows/ file and no shared fixture. */

#include <stdio.h>
#include <stdint.h>
#include <string.h>

#include "V1Table.h"

/* totality, shared by both refusals: nothing counted, nothing written */
static int check_nothing_moved( const Cell * back, const TableReport * r )
{
    int64_t i;
    if ( r->clamped || r->unknown || r->kind_mismatch || r->widened || r->duplicate || r->retained || r->retain_lost )
    {
        printf( "F4 layout_malformed, truncated: a counter moved: clamped=%d unknown=%d kind_mismatch=%d widened=%d duplicate=%d retained=%d retain_lost=%d\n",
                r->clamped, r->unknown, r->kind_mismatch, r->widened, r->duplicate, r->retained, r->retain_lost );
        return 0;
    }
    for ( i = 0; i < (int64_t) sizeof( *back ); ++i )
    {
        if ( ( (const uint8_t *) back )[i] != 0x5A )
        {
            printf( "F4 layout_malformed, truncated: byte %lld is 0x%02x -- the refusal wrote a destination byte\n",
                    (long long) i, ( (const uint8_t *) back )[i] );
            return 0;
        }
    }
    return 1;
}

int main( void )
{
    Cell source;
    Cell back;
    TableReport r;
    uint8_t wire[256];
    uint8_t tiny[20];
    int64_t bytes, n;

    cell_reset( &source );
    source.power = 7;
    memcpy( source.label, "hello", 6 );
    source.label_length = 5;

    bytes = cell_fixed_measure( 1 );
    if ( bytes < 0 || bytes > (int64_t) sizeof( wire ) )
    {
        printf( "F4 layout_malformed, truncated: measure out of range: %lld\n", (long long) bytes );
        return 1;
    }
    if ( cell_fixed_save( &source, 1, wire, bytes ) != bytes )
    {
        printf( "F4 layout_malformed, truncated: the clean save failed\n" );
        return 1;
    }

    /* THE NEGATIVE CONTROL, FIRST: the uncut file reads, and reads back the
       value it saved, so every refusal below is the ONE break and not the
       file. */
    memset( &back, 0x5A, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );
    n = cell_fixed_load( &back, 1, wire, bytes, NULL, 0, NULL, &r );
    if ( n != 1 || r.refused || r.malformed )
    {
        printf( "F4 layout_malformed, truncated: the uncut file did not read clean: n=%lld refused=%d malformed=%d\n",
                (long long) n, r.refused, r.malformed );
        return 1;
    }
    if ( back.power != 7 || back.label_length != 5 || memcmp( back.label, "hello", 6 ) != 0 )
    {
        printf( "F4 layout_malformed, truncated: the uncut read lost the value: power=%d label_length=%d\n",
                (int) back.power, (int) back.label_length );
        return 1;
    }

    /* THE TRUNCATED LAYOUT: the file ends one byte inside the layout, so
       20 + L > bytes, the record gone with the cut. The NAME, not the flag. */
    memset( &back, 0x5A, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );
    n = cell_fixed_load( &back, 1, wire, kTableFixedHeaderBytes + 4 + cell_fixed_layout_bytes - 1, NULL, 0, NULL, &r );
    if ( n != -1 )
    {
        printf( "F4 layout_malformed, truncated: a layout cut to 20+L-1 bytes did not refuse: n=%lld\n", (long long) n );
        return 1;
    }
    if ( !r.refused || r.reason != SCHEMA_TABLE_LAYOUT_MALFORMED )
    {
        printf( "F4 layout_malformed, truncated: refused=%d reason=%d, not layout_malformed\n", r.refused, r.reason );
        return 1;
    }
    if ( r.malformed )
    {
        printf( "F4 layout_malformed, truncated: malformed set -- a refusal by name is not the residue\n" );
        return 1;
    }
    if ( !check_nothing_moved( &back, &r ) ) { return 1; }

    /* THE MINIMAL VECTOR FROM THE LAW, AND THE ORDER: twenty bytes that are
       all header, claiming L = 1, under a hash in no lineage. 20 + 1 > 20,
       and the length is read (step 3) before the hash select (steps 5 and
       6), so the answer is layout_malformed and not layout_newer. */
    memset( tiny, 0, sizeof( tiny ) );
    tiny[0] = kTableFixedForm;
    table_fixed_put64( tiny + kTableFixedHashAt, 0xDEADBEEFCAFEF00Dull );
    table_fixed_put32( tiny + kTableFixedHeaderBytes, 1u );
    memset( &back, 0x5A, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );
    n = cell_fixed_load( &back, 1, tiny, (int64_t) sizeof( tiny ), NULL, 0, NULL, &r );
    if ( n != -1 )
    {
        printf( "F4 layout_malformed, truncated: a 20-byte header claiming L=1 did not refuse: n=%lld\n", (long long) n );
        return 1;
    }
    if ( !r.refused || r.reason != SCHEMA_TABLE_LAYOUT_MALFORMED )
    {
        printf( "F4 layout_malformed, truncated: refused=%d reason=%d, not layout_malformed before the hash select\n",
                r.refused, r.reason );
        return 1;
    }
    if ( r.malformed )
    {
        printf( "F4 layout_malformed, truncated: malformed set -- a refusal by name is not the residue\n" );
        return 1;
    }
    if ( !check_nothing_moved( &back, &r ) ) { return 1; }

    /* THE BOUNDARY: 20 + L == bytes is NOT truncated -- the comparison is
       `>`, so a file that ends exactly at the layout's end reads zero
       records, no refusal, no flag. */
    memset( &back, 0x5A, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );
    n = cell_fixed_load( &back, 1, wire, kTableFixedHeaderBytes + 4 + cell_fixed_layout_bytes, NULL, 0, NULL, &r );
    if ( n != 0 || r.refused || r.malformed )
    {
        printf( "F4 layout_malformed, truncated: exactly 20+L bytes is not truncated: n=%lld refused=%d malformed=%d\n",
                (long long) n, r.refused, r.malformed );
        return 1;
    }

    /* THE DISTINCTION FROM F7: a file one byte UNDER the twenty-byte header
       is the malformed FLAG, never the layout_malformed NAME -- a file that
       short cannot even carry the length field that would lie. */
    memset( &back, 0x5A, sizeof( back ) );
    memset( &r, 0, sizeof( r ) );
    n = cell_fixed_load( &back, 1, wire, kTableFixedHeaderBytes + 4 - 1, NULL, 0, NULL, &r );
    if ( n != -1 )
    {
        printf( "F4 layout_malformed, truncated: a 19-byte file did not refuse: n=%lld\n", (long long) n );
        return 1;
    }
    if ( !r.malformed || r.refused )
    {
        printf( "F4 layout_malformed, truncated: 19 bytes is the malformed flag, not the name: malformed=%d refused=%d\n",
                r.malformed, r.refused );
        return 1;
    }
    if ( !check_nothing_moved( &back, &r ) ) { return 1; }

    printf( "F4 layout_malformed, truncated: a file cut inside its layout refuses by the name; 20+L exactly reads; under 20 is the flag\n" );
    return 0;
}
